src/plugins/stake-plugins/generic/policyAdapters/TarotPoolAdaptor.ts

```diff
@@ -82,17 +95,67 @@ export const makeTarotPoolAdapter = (policyConfig: StakePolicyConfig<TarotPoolAd
   } = adapterConfig
   const provider = new ethers.providers.FallbackProvider(rpcProviderUrls.map(url => new ethers.providers.JsonRpcProvider(url)))
 
+  // L1 fee utils ported from edge-currency-accountbased:
+  async function updateL1RollupParams(): Promise<void> {
+    try {
+      const oracleContract = OptimismFeeOracle__factory.connect(l1RollupParams.proxyContract, provider)
+      const scalar = await oracleContract.scalar()
+      const feeRes = await oracleContract.l1BaseFee()
+      const bigNumberSafeGasPriceMultiplier = mul(l1RollupParams.maxGasPriceL1Multiplier, '100')
+      const bigNumberSafeGasPriceDivisor = '100'
+
+      l1RollupParams = {
+        ...l1RollupParams,
+        dynamicOverhead: scalar.toString(),
+        gasPriceL1Wei: feeRes.mul(bigNumberSafeGasPriceMultiplier).div(bigNumberSafeGasPriceDivisor).toString()
+      }
+    } catch (e: any) {
+      console.warn('makeTarotPoolAdapter Failed to update L1RollupParams', e)
+    }
+  }
+  const calcL1RollupFees = (populatedTx: ethers.PopulatedTransaction): string => {
+    const txCopy = { ...populatedTx }
+    delete txCopy.customData
+
+    const unsignedRawTxData = ethers.utils.serializeTransaction(txCopy)
+    const unsignedRawTxBytesArray = unsignedRawTxData.match(/(.{1,2})/g)
+    if (unsignedRawTxBytesArray == null) {
+      throw new Error('Invalid rawTx string')
+    }
+
+    let rawTxCost = 0
+    for (let i = 0; i < unsignedRawTxBytesArray.length; i++) {
+      if (unsignedRawTxBytesArray[i] === '00') {
+        rawTxCost += 4 // cost for zero byte
+      } else {
+        rawTxCost += 16 // cost for non-zero byte
+      }
+    }
+
+    const { dynamicOverhead, fixedOverhead, gasPriceL1Wei } = l1RollupParams
+    const MAX_SIGNATURE_COST = '1040' // (32 + 32 + 1) * 16 max cost for adding r, s, v signatures to raw transaction
+
+    const gasUsed = add(add(rawTxCost.toString(), fixedOverhead), MAX_SIGNATURE_COST)
+    const scalar = div(dynamicOverhead, '1000000', 18)
+    const total = ceil(mul(mul(gasPriceL1Wei, gasUsed), scalar), 0)
+
+    return total
+  }
```

Comment:
This looks familiar. I'm going to say it's safe to assume this is copy pasta code from accountbased and so therefore I don't have to review too hard?

Reply:
Yup it's ported from accountbased